sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuApproximatePercentile.scala

```diff
@@ -0,0 +1,147 @@
+  protected final lazy val outputBuf: AttributeReference =
+    inputBuf.copy("outputBuf", CudfTDigest.dataType)(inputBuf.exprId, inputBuf.qualifier)
+
+  /**
+   * Using child references, define the shape of the vectors sent to the window operations
+   */
+  override val windowInputProjection: Seq[Expression] = Seq.empty
+
+  /**
+   * Create the aggregation operation to perform for Windowing. The input to this method
+   * is a sequence of (index, ColumnVector) that corresponds one to one with what was
+   * returned by [[windowInputProjection]].  The index is the index into the Table for the
+   * corresponding ColumnVector. Some aggregations need extra values.
+   */
+  override def windowAggregation(inputs: Seq[(ColumnVector, Int)]): RollingAggregationOnColumn = {
+    throw new UnsupportedOperationException()
+  }
+
+  override def nullable: Boolean = false
+
+  // Mark as lazy so that percentageExpression is not evaluated during tree transformation.
+  private lazy val (returnPercentileArray, percentiles) =
+    percentageExpression match {
+      case GpuLiteral(v, _) => makeArray(v)
+      case _ => makeArray(percentageExpression.eval())
+    }
+
+  private def makeArray(v: Any)  = v match {
+    // Rule ImplicitTypeCasts can cast other numeric types to double
+    case null => (false, null)
+    case num: Double => (false, Array(num))
+    case arrayData: ArrayData => (true, arrayData.toDoubleArray())
+  }
+
+
+  // The result type is the same as the input type.
+  private lazy val internalDataType: DataType = {
+    if (returnPercentileArray) ArrayType(child.dataType,
+      containsNull = false) else child.dataType
+  }
+
+  override def dataType: DataType = internalDataType
+
+  override def children: Seq[Expression] = Seq(child, percentageExpression, accuracyExpression)
+}
+
+object GpuApproximatePercentile {
+  def apply(child: Expression, percentageExpression: GpuLiteral): GpuApproximatePercentile = {
+    GpuApproximatePercentile(child, percentageExpression,
+      GpuLiteral(ApproximatePercentile.DEFAULT_PERCENTILE_ACCURACY))
+  }
+}
+
+case class ApproxPercentileFromTDigestExpr(
+    child: Expression,
+    percentiles: Array[Double],
+    finalDataType: DataType)
+  extends GpuExpression {
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    val expr = child.asInstanceOf[GpuExpression]
+    withResource(expr.columnarEval(batch).asInstanceOf[GpuColumnVector]) { cv =>
+      withResource(cv.getBase.approxPercentile(percentiles)) { percentiles =>
+        if (finalDataType == DataTypes.DoubleType) {
+          GpuColumnVector.from(percentiles.incRefCount(), dataType)
+        } else {
+          // cast cuDF Array[Double] to Array[finalDataType]
+          withResource(percentiles.getChildColumnView(0)) { childView =>
+            withResource(recursiveDoColumnar(childView, DataTypes.DoubleType, finalDataType,
+                ansiMode = SQLConf.get.ansiEnabled, legacyCastToString = false)) { childCv =>
```

Comment:
I think ansiMode is always going to be false here. The original spark code didn't check ansiMode at all, and in theory we could just cast it ourselves with relatively little effort for everything except DecimalType which we still are not going to do great at, but at least when we do "fix" this in cast it would also be fixed here.

```
    val result = child.dataType match {
      case DateType => doubleResult.map(_.toInt)
      case TimestampType => doubleResult.map(_.toLong)
      case ByteType => doubleResult.map(_.toByte)
      case ShortType => doubleResult.map(_.toShort)
      case IntegerType => doubleResult.map(_.toInt)
      case LongType => doubleResult.map(_.toLong)
      case FloatType => doubleResult.map(_.toFloat)
      case DoubleType => doubleResult
      case _: DecimalType => doubleResult.map(Decimal(_))
      case other: DataType =>
        throw new UnsupportedOperationException(s"Unexpected data type ${other.catalogString}")
    }
```